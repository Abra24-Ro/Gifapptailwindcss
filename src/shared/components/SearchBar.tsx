import { useState, useEffect } from "react";
import type { KeyboardEvent } from "react";


interface Props {
  placeholder?: string;
  onQuery: (query: string) => void;
}

export const SearchBar = ({ placeholder, onQuery }: Props) => {
  const [query, setQuery] = useState("");

  const handleSubmit = () => {
    onQuery(query);
    setQuery("");
  };

useEffect(() => {
  if (query === "") {
    onQuery("");
  }
}, [query]); // ðŸ‘ˆ solo query


  const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter" && query.trim() !== "") {
      handleSubmit();
    }
  };

  return (
    <section className="max-w-3xl mx-auto px-6 py-6">
      <div className="flex gap-2">
        <input
          value={query}
          onChange={(e) => setQuery(e.target.value)}
          type="text"
          placeholder={placeholder || "Buscar gifs"}
          className="flex-1 px-3 py-2 text-sm border border-gray-300 rounded-md bg-white placeholder-gray-400 focus:border-blue-500 focus:ring-1 focus:ring-blue-500 outline-none"
          onKeyDown={handleKeyDown}
        />
        <button
          className="px-4 py-2 text-sm font-medium bg-gray-900 text-white rounded-md hover:bg-gray-800 transition-colors"
          onClick={handleSubmit}
        >
          Buscar
        </button>
      </div>
    </section>
  );
};
