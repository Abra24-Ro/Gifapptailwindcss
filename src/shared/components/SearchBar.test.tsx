import { describe, expect, test, vi } from "vitest";
import { SearchBar } from "./SearchBar";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

describe("SearchBar", () => {
  test("should render searchbar properly", () => {
    const { container } = render(<SearchBar onQuery={() => {}} />);

    expect(container).toMatchSnapshot();
    expect(screen.getByRole("textbox")).toBeTruthy();
  });

  test("should call Onquery with the correct value after 700ms", async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "Goku" } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledWith("Goku");
      expect(onQuery).toHaveBeenCalledTimes(1);
    });
  });

  test("should call only one with the last value (debounce)", async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "t" } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledWith("t");
      expect(onQuery).toHaveBeenCalledTimes(1);
    });
  });

  test("should call Onquery when clicked with the input value", () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "t" } });

    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(onQuery).toHaveBeenCalledWith("t");
    expect(onQuery).toHaveBeenCalledTimes(1);
  });

  test("should the input has the correct placeHolader value", () => {
    const placeHolder = "Busca lo que quieras";

    render(<SearchBar onQuery={() => {}} placeholder={placeHolder} />);
    expect(screen.getByPlaceholderText(placeHolder)).toBeTruthy();
    
  });
});
