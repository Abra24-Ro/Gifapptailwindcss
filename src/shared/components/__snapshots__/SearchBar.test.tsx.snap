// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render searchbar properly 1`] = `
<div>
  <section
    class="max-w-3xl mx-auto px-6 py-6"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="flex-1 px-3 py-2 text-sm border border-gray-300 rounded-md bg-white placeholder-gray-400 focus:border-blue-500 focus:ring-1 focus:ring-blue-500 outline-none"
        placeholder="Buscar gifs"
        type="text"
        value=""
      />
      <button
        class="px-4 py-2 text-sm font-medium bg-gray-900 text-white rounded-md hover:bg-gray-800 transition-colors"
      >
        Buscar
      </button>
    </div>
  </section>
</div>
`;
